olive/common/ort_inference.py

```diff
@@ -51,26 +75,299 @@ def get_ort_inference_session(
         for key, value in extra_session_config.items():
             sess_options.add_session_config_entry(key, value)
 
-    if isinstance(execution_provider, str):
-        execution_provider = [execution_provider]
-    else:
-        # execution providers should be list[str]
-        assert isinstance(execution_provider, list)
-        assert all(isinstance(ep, str) for ep in execution_provider)
-
-    for idx, ep in enumerate(execution_provider):
-        if ep == "QNNExecutionProvider":
+    # execution providers and provider options
+    providers, provider_options = check_and_normalize_provider_args(
+        inference_settings.get("execution_provider"),
+        inference_settings.get("provider_options"),
+        ort.get_available_providers(),
+    )
+    for idx, provider in enumerate(providers):
+        if provider in ["CUDAExecutionProvider", "DmlExecutionProvider"] and device_id is not None:
+            provider_options[idx]["device_id"] = str(device_id)
+        elif provider == "QNNExecutionProvider":
             # add backend_path for QNNExecutionProvider
-            execution_provider[idx] = ("QNNExecutionProvider", {"backend_path": "QnnHtp.dll"})
-            break
+            provider_options[idx]["backend_path"] = "QnnHtp.dll"
+    logger.debug(f"Normalized providers: {providers}, provider_options: {provider_options}")
 
     # dml specific settings
-    if len(execution_provider) >= 1 and execution_provider[0] == "DmlExecutionProvider":
+    if len(providers) >= 1 and providers[0] == "DmlExecutionProvider":
```

Comment:
What if the order of EP provided by user is not `DmlEP, CPUEP...`? How about:
```
if 'DmlExecutionProvider' in providers
```

Reply:
This is the pre-existing logic so I didn't want to change the behavior in the PR https://github.com/microsoft/Olive/blob/683ed65a5c11c614759d186c607bcc26553afae3/olive/common/ort_inference.py#L68